### PR TITLE
Message request UI

### PR DESCRIPTION
**Pull request by warrickct**
Message Requests with updated functionality.

### CODE REVIEW

File: stylesheets/_session.scss [hunk 1]
```diff
@@ -219,14 +219,21 @@ textarea {
   & > *:hover:not(svg) {
     filter: brightness(80%);
   }
+
+  &.simple {
+    &.danger {
+      color: var(--color-destructive);
+    }
+  }
```
Comment (Bilb) [inline stylesheets/_session.scss:227]: is this used anywhere?

File: stylesheets/_session.scss [hunk 2]
```diff
@@ -219,14 +219,21 @@ textarea {
...
 }
 
 .session-label {
   color: $session-color-white;
   padding: $session-margin-sm;
   width: 100%;
   border-radius: 2px;
-  text-align: center;
+  text-align: centere
```
Comment (Bilb) [inline stylesheets/_session.scss:235]: typo I guess

File: ts/components/conversation/SessionConversation.tsx [hunk 3]
```diff
@@ -238,10 +277,34 @@ export class SessionConversation extends React.Component<Props, State> {
...
+                    onClick={handleDeclineConversationRequest}
+                  />
+                </div>
+                {/* 
+                  Disabling for now until report as spam is added in
+                <SessionButton
+                  buttonColor={SessionButtonColor.Danger}
+                  buttonType={SessionButtonType.Simple}
+                  text={window.i18n('reportAsSpam')}
+                /> */}
```
Comment (Bilb) [inline ts/components/conversation/SessionConversation.tsx:305]: you can remove that  for now.

File: ts/components/conversation/composition/CompositionBox.tsx [hunk 4]
```diff
@@ -424,16 +424,27 @@ class CompositionBoxInner extends React.Component<Props, State> {
...
+        ? i18n('messageRequestPending')
+        : isKickedFromGroup
+        ? i18n('youGotKickedFromGroup')
+        : left
+        ? i18n('youLeftTheGroup')
+        : isBlocked && isPrivate
+        ? i18n('unblockToSend')
+        : isBlocked && !isPrivate
+        ? i18n('unblockGroupToSend')
+        : i18n('sendMessage');
```
Comment (Bilb) [inline ts/components/conversation/composition/CompositionBox.tsx:447]: can you move this to a function with some real ifs and some comments between each steps so later update are easier/

File: ts/components/conversation/message/message-item/MessageRequestResponse.tsx [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  if (conversationId) {
+    profileName =
+      getConversationController()
+        .get(conversationId)
+        .getProfileName() + '';
+  }
+  const msgText =
+    profileName && props.source === UserUtils.getOurPubKeyStrFromCache()
+      ? window.i18n('messageRequestAcceptedOurs', [profileName])
+      : window.i18n('messageRequestAccepted');
```
Comment (Bilb) [inline ts/components/conversation/message/message-item/MessageRequestResponse.tsx:22]: can you use one of the selector or hook to fetch this profileName instead? It's just that set like that, this won't be updated if we set the nickname or if the profileName of that user changes before a refresh of the conversation.

Something like useConversationUsername() (search for it to see how it is used)

File: ts/components/conversation/SessionConversation.tsx [hunk 6]
```diff
@@ -238,10 +277,34 @@ export class SessionConversation extends React.Component<Props, State> {
           <div className={classNames('conversation-info-panel', showMessageDetails && 'show')}>
             <MessageDetail />
           </div>
-
           {lightBoxOptions?.media && this.renderLightBox(lightBoxOptions)}
 
           <div className="conversation-messages">
+            {showMsgRequestUI && (
+              <ConversationRequestBanner>
+                <div className="conversation-request-banner__row">
```
Comment (Bilb) [inline ts/components/conversation/SessionConversation.tsx:285]: why don't you use a styled-component here too, instead of a normal div here and some css below at line 601?
```
.conversation-request-banner__row {
    display: flex;
    flex-direction: row;
    gap: var(--margins-lg);
    justify-content: center;
  }
```

File: ts/components/conversation/SessionConversation.tsx [hunk 7]
```diff
@@ -221,7 +228,39 @@ export class SessionConversation extends React.Component<Props, State> {
...
+      window.inboxStore?.dispatch(
+        updateConfirmModal({
+          okText: window.i18n('decline'),
+          cancelText: window.i18n('cancel'),
+          message: window.i18n('declineRequestMessage'),
+          onClickOk: async () => {
+            declineConversation(selectedConversation.id, false);
+            blockConvoById(selectedConversation.id);
+            forceSyncConfigurationNowIfNeeded();
+          },
```
Comment (Bilb) [inline ts/components/conversation/SessionConversation.tsx:249]: I think you should await each of these calls here, as the last one might not take an uptodate value to do the sync otherwise.

File: ts/components/conversation/SessionConversation.tsx [hunk 8]
```diff
@@ -221,7 +228,39 @@ export class SessionConversation extends React.Component<Props, State> {
       return <MessageView />;
     }
 
+    const conversation = getConversationController().get(selectedConversation.id);
```
Comment (Bilb) [inline ts/components/conversation/SessionConversation.tsx:231]: why don't you rely on `this.props.selectedConversation` directly?

File: ts/state/ducks/SessionTheme.tsx [hunk 9]
```diff
@@ -138,6 +144,23 @@ export const switchHtmlToDarkTheme = () => {
...
+    darkMessageRequestBannerIconBackground
+  );
+  document.documentElement.style.setProperty(
+    '--color-request-banner-unread-background',
+    darkMessageRequestBannerUnreadBackground
+  );
+  document.documentElement.style.setProperty(
+    '--color-request-banner-icon',
+    darkMessageRequestBannerIcon
+  );
```
Comment (Bilb) [inline ts/state/ducks/SessionTheme.tsx:163]: are those really new colors ?
If not, be sure to rely on existing values instead. we don't want to add too many colors to the body element

File: ts/state/ducks/SessionTheme.tsx [hunk 10]
```diff
@@ -276,6 +304,22 @@ export const switchHtmlToLightTheme = () => {
...
+    lightMessageRequestBannerIconBackground
+  );
+  document.documentElement.style.setProperty(
+    '--color-request-banner-unread-background',
+    lightMessageRequestBannerUnreadBackground
+  );
+  document.documentElement.style.setProperty(
+    '--color-request-banner-icon',
+    lightMessageRequestBannerIcon
+  );
```
Comment (Bilb) [inline ts/state/ducks/SessionTheme.tsx:322]: same

File: ts/session/sending/MessageQueue.ts [hunk 11]
```diff
@@ -32,7 +33,8 @@ type ClosedGroupMessageType =
   | ExpirationTimerUpdateMessage
   | ClosedGroupEncryptionPairMessage
   | UnsendMessage
-  | ClosedGroupEncryptionPairRequestMessage;
+  | ClosedGroupEncryptionPairRequestMessage
+  | MessageRequestResponse;
```
Comment (Bilb) [inline ts/session/sending/MessageQueue.ts:37]: why is this needed? 
This shouldn't be a message sent to a closed group right?

File: ts/models/conversation.ts [hunk 12]
```diff
@@ -705,6 +731,37 @@ export class ConversationModel extends Backbone.Model<ConversationAttributes> {
...
+    // const msgRequestResponseParams:
+    const timestamp = Date.now();
+
+    const messageRequestResponseParams = {
+      timestamp,
+      publicKey,
+      isApproved,
+    };
+
+    const messageRequestResponse = new MessageRequestResponse(messageRequestResponseParams);
```
Comment (Bilb) [inline ts/models/conversation.ts:745]: can you move that `fromHexToArray` to the `MessageRequestResponse` constructor instead? And just pass the string as is

File: ts/interactions/conversationInteractions.ts [hunk 13]
```diff
@@ -121,20 +121,48 @@ export async function unblockConvoById(conversationId: string) {
...
+  if (!convoToApprove || convoToApprove.isApproved()) {
     window?.log?.info('Conversation is already approved.');
     return;
   }
 
-  await conversationToApprove.setIsApproved(true);
+  Promise.all([
+    await convoToApprove.setIsApproved(true),
+    await convoToApprove.setDidApproveMe(true),
+  ]);
```
Comment (Bilb) [inline ts/interactions/conversationInteractions.ts:137]: each of those calls do a commit() to the db. Might be a good idea to have another conversationModel function to update those two fields when needed with only one commit() instead

File: ts/models/conversation.ts [hunk 14]
```diff
@@ -634,11 +643,28 @@ export class ConversationModel extends Backbone.Model<ConversationAttributes> {
...
+          void forceSyncConfigurationNowIfNeeded();
+        }
+        // void forceSyncConfigurationNowIfNeeded();
+      }
+
+      // TODO: remove once dev-tested
+      if (chatMessageParams.body?.includes('unapprove')) {
+        await this.setIsApproved(false);
+        await this.setDidApproveMe(false);
+        // void forceSyncConfigurationNowIfNeeded();
```
Comment (Bilb) [inline ts/models/conversation.ts:667]: to remove

File: app/sql.js [hunk 15]
```diff
@@ -1335,6 +1336,55 @@ function updateToLokiSchemaVersion19(currentVersion, db) {
...
+      .prepare(
+        `
+        SELECT json FROM ${CONVERSATIONS_TABLE} WHERE
+        type = 'group' AND
+        id NOT LIKE 'publicChat:%';
+      `
+      )
+      .all();
+
+    console.warn({ closedGroupRows });
```
Comment (Bilb) [inline app/sql.js:1363]: avoid using console.warn. this should be used for debugging only. You can have console.info if you want to

File: app/sql.js [hunk 16]
```diff
@@ -1335,6 +1336,55 @@ function updateToLokiSchemaVersion19(currentVersion, db) {
...
+    // all closed group admins
+    const closedGroupRows = db
+      .prepare(
+        `
+        SELECT json FROM ${CONVERSATIONS_TABLE} WHERE
+        type = 'group' AND
+        id NOT LIKE 'publicChat:%';
+      `
+      )
+      .all();
```
Comment (Bilb) [inline app/sql.js:1361]: you can call `getAllClosedGroupConversationsV1` after renaming it to some more meaningful (like without the v1)

File: app/sql.js [hunk 17]
```diff
@@ -1335,6 +1336,55 @@ function updateToLokiSchemaVersion19(currentVersion, db) {
...
+
+    console.warn({ closedGroupRows });
+
+    const adminIds = closedGroupRows.map(json => {
+      return jsonToObject(json).groupAdmins;
+    });
+    console.warn({ adminIds });
```
Comment (Bilb) [inline app/sql.js:1368]: same

File: ts/models/conversation.ts [hunk 18]
```diff
@@ -634,11 +643,28 @@ export class ConversationModel extends Backbone.Model<ConversationAttributes> {
         lokiProfile: UserUtils.getOurProfile(),
       };
 
-      const updateApprovalNeeded =
-        !this.isApproved() && (this.isPrivate() || this.isMediumGroup() || this.isClosedGroup());
-      if (updateApprovalNeeded) {
+      const shouldApprove = !this.isApproved() && this.isPrivate();
+      const hasMsgsFromOther =
+        (await getMessagesByConversation(this.id, { type: MessageDirection.incoming })).length > 0;
+      console.warn(hasMsgsFromOther);
```
Comment (Bilb) [inline ts/models/conversation.ts:649]: to remove

File: ts/models/conversation.ts [hunk 19]
```diff
@@ -634,11 +643,28 @@ export class ConversationModel extends Backbone.Model<ConversationAttributes> {
...
+      if (shouldApprove) {
         await this.setIsApproved(true);
-        void forceSyncConfigurationNowIfNeeded();
+        if (!this.didApproveMe() && hasMsgsFromOther) {
+          console.warn('This is a reply message sending message request acceptance response.');
```
Comment (Bilb) [inline ts/models/conversation.ts:653]: make this a window.log.info

File: ts/receiver/closedGroups.ts [hunk 20]
```diff
@@ -111,6 +111,18 @@ export async function handleClosedGroupControlMessage(
...
+    if (
+      getConversationController()
+        .get(envelope.senderIdentity)
+        .isApproved() == false
+    ) {
+      window?.log?.info(
+        'Received new closed group message from an unapproved sender -- dropping message.'
+      );
+      // TODO: remove console output
+      console.warn('Received unapproved closed group invite msg');
```
Comment (Bilb) [inline ts/receiver/closedGroups.ts:123]: same

File: ts/receiver/closedGroups.ts [hunk 21]
```diff
@@ -111,6 +111,18 @@ export async function handleClosedGroupControlMessage(
     return;
   }
   if (type === Type.NEW) {
+    if (
+      getConversationController()
+        .get(envelope.senderIdentity)
+        .isApproved() == false
+    ) {
```
Comment (Bilb) [inline ts/receiver/closedGroups.ts:118]: any reasons for the loose equaliity check?

File: ts/receiver/queuedJob.ts [hunk 22]
```diff
@@ -223,10 +223,10 @@ async function handleRegularMessage(
...
     await handleSyncedReceipts(message, conversation);
 
-    if (window.lokiFeatureFlags.useMessageRequests) {
-      // assumes sync receipts are always from linked device outgoings
-      await conversation.setIsApproved(true);
-    }
+    // if (window.lokiFeatureFlags.useMessageRequests) {
+    //   // assumes sync receipts are always from linked device outgoings
+    //   await conversation.setIsApproved(true);
+    // }
```
Comment (Bilb) [inline ts/receiver/queuedJob.ts:229]: do we need to keep this or not?

File: ts/components/conversation/SessionConversation.tsx (hunk 8 above)
Comment (warrickct) [reply ts/components/conversation/SessionConversation.tsx:231]: fingers moving faster than the brain
